spec/src/consensus.rs

```diff
@@ -93,59 +100,56 @@ impl ProposalWindow {
-    pub id: String,
-    pub genesis_block: BlockView,
-    pub genesis_hash: Byte32,
-    pub epoch_reward: Capacity,
-    pub secondary_epoch_reward: Capacity,
-    pub max_uncles_num: usize,
-    pub orphan_rate_target: RationalU256,
-    pub epoch_duration_target: u64,
-    pub tx_proposal_window: ProposalWindow,
-    pub proposer_reward_ratio: Ratio,
-    pub pow: Pow,
-    // For each input, if the referenced output transaction is cellbase,
-    // it must have at least `cellbase_maturity` confirmations;
-    // else reject this transaction.
-    pub cellbase_maturity: BlockNumber,
-    // This parameter indicates the count of past blocks used in the median time calculation
-    pub median_time_block_count: usize,
-    // Maximum cycles that all the scripts in all the commit transactions can take
-    pub max_block_cycles: Cycle,
-    // Maximum number of bytes to use for the entire block
-    pub max_block_bytes: u64,
-    // block version number supported
-    pub block_version: Version,
-    // block version number supported
-    pub max_block_proposals_limit: u64,
-    pub genesis_epoch_ext: EpochExt,
+pub struct ConsensusBuilder {
+    id: String,
+    genesis_block: BlockView,
+    genesis_hash: Byte32,
+    epoch_reward: Capacity,
+    secondary_epoch_reward: Capacity,
+    max_uncles_num: usize,
+    orphan_rate_target: RationalU256,
+    epoch_duration_target: u64,
+    tx_proposal_window: ProposalWindow,
+    proposer_reward_ratio: Ratio,
+    pow: Pow,
+    cellbase_maturity: BlockNumber,
+    median_time_block_count: usize,
+    max_block_cycles: Cycle,
+    max_block_bytes: u64,
+    block_version: Version,
+    max_block_proposals_limit: u64,
+    genesis_epoch_ext: EpochExt,
+    satoshi_pubkey_hash: H160,
+    satoshi_cell_occupied_ratio: Ratio,
 }
 
 // genesis difficulty should not be zero
-impl Default for Consensus {
+impl Default for ConsensusBuilder {
     fn default() -> Self {
         let input = CellInput::new_cellbase_input(0);
         let witness = Script::default().into_witness();
         let cellbase = TransactionBuilder::default()
             .input(input)
             .witness(witness)
             .build();
-        let dao = genesis_dao_data(vec![&cellbase]).unwrap();
+        let dao = genesis_dao_data_with_satoshi_gift(
+            vec![&cellbase],
+            &SATOSHI_PUBKEY_HASH,
+            SATOSHI_CELL_OCCUPIED_RATIO,
+        )
+        .unwrap();
         let genesis_block = BlockBuilder::default()
             .difficulty(U256::one().pack())
             .dao(dao)
             .transaction(cellbase)
             .build();
 
-        Consensus::new(genesis_block, DEFAULT_EPOCH_REWARD)
+        ConsensusBuilder::new(genesis_block, DEFAULT_EPOCH_REWARD)
     }
 }
 
-#[allow(clippy::op_ref)]
-impl Consensus {
-    pub fn new(genesis_block: BlockView, epoch_reward: Capacity) -> Consensus {
+impl ConsensusBuilder {
+    pub fn new(genesis_block: BlockView, epoch_reward: Capacity) -> Self {
```

Comment:
I think this `new(..)` method has side effects.

Remove Side Effects:
- Please use `ConsensusBuilder::default().set(..)....set(..).build(..)` to rewrite it, if any parameters we have to set, then these parameters should be the arguments of `new(..)`.

Reply:
I don't think the `initialize` is `side effects`.

Reply:
May be I did not say clearly, or I used inappropriate words.

I mean: `T::new(inputs)` should equal to `T::default().update(inputs)`, not `T::default().update(inputs).update_some_extras_things()`, so the function signature can show what does the function do, clearly.

Reply:
If new is just 1 to 1 equals to the builder, why we need the `new` function? This function is of course surely to put some extra initialize logic.